### ISSUE TITLE
CRITICAL: No causal attention mask — bidirectional attention in autoregressive training

### ISSUE THREAD

**Issue by noahgift** [paiml/entrenar#310]
## Problem

`CudaTransformerBlock::compute_attention_cuda()` applies softmax to the full
[num_heads, seq, seq] attention score matrix without a causal mask. Every token
attends to every other token, including future tokens.

This is a **bidirectional** attention pattern being used in **autoregressive**
(causal LM) training. The model sees future tokens during forward pass, making
gradients less useful for learning autoregressive prediction.

## Evidence

`cuda_block.rs` lines 855-875:
```rust
// Step 6: Row-wise softmax → attn_weights [num_heads * seq, seq] (in-place)
batched_softmax_forward(
    &scores_view,
    &mut self.scratch.attn_scores,
    total_rows,
    seq,
    stream,
)?;
```

No causal mask applied before softmax. All positions attend to all positions.

## Impact

Combined with entrenar#309 (sinusoidal init, now FIXED):
- PyTorch (causal attention + normal init): val_ppl=50 at step 1K
- Entrenar v17 (normal init, NO causal mask): val_ppl=729 at step 1K
- **14.6x convergence gap** remains after init fix

## Literature

- [Vaswani et al. (2017)](https://arxiv.org/abs/1706.03762) — "Attention Is All
  You Need" §3.1: decoder self-attention uses masking to prevent positions from
  attending to subsequent positions
- [Dao et al. (2022)](https://arxiv.org/abs/2205.14135) — FlashAttention: fused
  causal mask with tiled softmax, 2x speedup from skipping masked tiles
- [PyTorch SDPA](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html)
  — `is_causal=True` creates `tril(ones(L,S)).bool()` mask, fills upper triangle
  with `-inf`

## Specification

Before softmax, apply:
```
scores[head, i, j] = -inf   for all j > i
```

This is equivalent to: `scores = scores + causal_bias` where
`causal_bias[i,j] = 0 if j <= i, -inf if j > i`.

### Options

1. **Separate kernel**: Apply `-inf` mask to upper triangle, then existing softmax
2. **Fused causal softmax**: New kernel that masks + softmax in one pass (trueno
   already has `AttentionKernel::with_causal()` in `trueno-gpu/src/kernels/attention/`)
3. **Use trueno's flash attention**: Replace the manual Q@K^T→scale→softmax→@V
   pipeline with `AttentionKernel::with_causal()` (biggest refactor but best perf)

## Contract

See `contracts/causal-attention-mask-v1.yaml` (C-CAUSAL-001).

## References

- entrenar#309 (init fix + this finding)
- [Vaswani et al. (2017)](https://arxiv.org/abs/1706.03762)
- [Dao et al. (2022)](https://arxiv.org/abs/2205.14135)
- albor PyTorch canary: `scripts/canary_pytorch.py`
